File: Top_Interview_Questions/Hard/44. Wildcard Matching/solution4.cpp
#include <string>
#include <vector>
using namespace std;

/**
 * 想法: 改用 dp table法來完成，為了符合先前的邏輯，所以會繼續使用 bottom-up 的
 * 方式，dp[i][j] 代表 s[0..i) 是否 match p[0..j)
 * 時間: O(m*n), 空間: O(m*n)
 */

class Solution
{
  public:
    bool isMatch(string s, string p)
    {
        int m = s.length(), n = p.length();
        vector<vector<bool>> dp(m + 1, vector<bool>(n + 1, false));
        dp[m][n] = true;
        for (int i = m; i >= 0; i--)
        {
            for (int j = n - 1; j >= 0; j--)
            {
                if (p[j] == '*')
                {
                    dp[i][j] = dp[i][j + 1] || (i < m && dp[i + 1][j]);
                }
                else
                {
                    dp[i][j] = i < m && (p[j] == s[i] || p[j] == '?') && dp[i + 1][j + 1];
                }
            }
        }
        return dp[0][0];
    }
};

File: Top_Interview_Questions/Hard/44. Wildcard Matching/solution4.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        dp = [[False for _ in range(n + 1)] for _ in range(m + 1)]
        dp[m][n] = True
        for i in range(m, -1, -1):
            for j in range(n - 1, -1, -1):
                if p[j] == "*":
                    dp[i][j] = dp[i][j + 1] or (i < m and dp[i + 1][j])
                else:
                    dp[i][j] = i < m and (p[j] == s[i] or p[j] == "?") and dp[i + 1][j + 1]
        return dp[0][0]
